### PR TITLE
resize-convolution instead transposed-convolution to avoid checkerboard artifacts

### PR DESCRIPTION
**Pull request by botcs**
See the reasoning behind the reimplementation:
https://distill.pub/2016/deconv-checkerboard/

I became aware of some checkerboard effects while training the CyclicGAN network:
http://i.imgur.com/uTgsC2l.png

because it resembled this: 
http://i.imgur.com/M8r47Bp.png

now it looks like this (during training on Van Gogh set):
http://i.imgur.com/wf7B9sT.png

### CODE REVIEW

**Comment by aman-tiwari** [comment]
Did you also try PixelShuffle?

**Comment by botcs** [comment]
I am now testing a transformer-net alternative with PixelShuffle using Fast-Neural-Style loss, CycleGAN still have some convergence issues, and the problem seems to be invariant of the capacity

**Comment by SsnL** [comment]
Thanks for your PR, @botcs !

I took a brief look at the code. This PR contains much more than what you have described. You should definitely clean up stuff like normalization, video, printing, etc. 

Some questions:
1. Why [change strided Conv to Conv+MaxPooling in downsampling](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/pull/64/files#diff-a56d77751e639c0dd4a453fc554f55dbR167)?
2. Why use bilinear interpolation? The post authors suggest nearest neighbor gives them best results. What's the reasoning behind your choice? And maybe make it an option?

By the way, nn.Upsample is available now. You may also want to uncomment [this line](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/pull/64/files#diff-a56d77751e639c0dd4a453fc554f55dbR186) and remove the now deprecated nn.UpsamplingBilinear2d? :)

**Comment by botcs** [comment]
Yeah, sorry for the inconvenience I will clean the code up in two weeks, and make another PR, with all the listed details - if that is acceptable for you

**Comment by beniroquai** [comment]
Is there any activity on this topic? I'm also interested to implement this feature - at least to get it work ;-)

**Comment by SsnL** [comment]
@beniroquai I have an implementation here: https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/issues/190

**Comment by beniroquai** [comment]
Wow that was quick! Thanks a lot. Did you plan to make a PR in this repo as well? So the only thing is to exchange the resnet code part, right? 
Will try it! Eager to see if this is solving my issues ;)

**Comment by SsnL** [comment]
@beniroquai that's right. Hope that it helps! :)

**Comment by beniroquai** [comment]
Thanks!  It's working way better now? Maybe stupid question, but Would it
make sense to do the same in the unet?

**Comment by SsnL** [comment]
@beniroquai Maybe, you can try adept the code to unet. It’d be cool if it
also helps unet. If you do some experiment, please let me know the results.
:)

**Comment by beniroquai** [comment]
Thanks for getting back! Would you suggest to exchange the convTranspose2D
or do I need to do anything else? I'm not that familiar with the coding in
pytorch though. I'll have a try! Thanks!

**Comment by mrgloom** [comment]
Seems nearest neighbour upsampling not solving all issues, but bring another one - random color edges:
https://distill.pub/2016/deconv-checkerboard/
https://twitter.com/ch402/status/793483428649766913?lang=en
https://distill.pub/2016/deconv-checkerboard/assets/style_clean.png
https://twitter.com/alexjc/status/788115081468997634

File: models/networks.py [hunk 1]
```diff
@@ -149,21 +163,40 @@ def __init__(self, input_nc, output_nc, ngf=64, norm_layer=nn.BatchNorm2d, use_d
         for i in range(n_downsampling):
             mult = 2**i
             model += [nn.Conv2d(ngf * mult, ngf * mult * 2, kernel_size=3,
-                                stride=2, padding=1),
+                                stride=1, padding=1),
+                      nn.MaxPool2d(2),
```
Comment (NataliaDiaz) [inline models/networks.py:167]: Hi! Why Pooling is used for avoiding checkerboard artifacts if
" Max pooling was previously linked to high-frequency artifacts in [12].)" https://distill.pub/2016/deconv-checkerboard/
